numcodecs/compat.py

```diff
@@ -27,80 +25,150 @@
-            buf = np.frombuffer(buf, dtype=out.dtype)
+    # copy memory
+    np.copyto(out, buf)
 
-        # ensure shapes are compatible
-        if buf.shape != out.shape:
-            if out.flags.f_contiguous:
-                order = 'F'
-            else:
-                order = 'C'
-            buf = buf.reshape(out.shape, order=order)
+    return out
 
-        # copy via numpy
-        np.copyto(out, buf)
 
-    # handle generic buffer destination
-    else:
+def ensure_text(l, encoding='utf-8'):
+    if isinstance(l, text_type):
+        return l
+    else:  # pragma: py3 no cover
+        return text_type(l, encoding=encoding)
 
-        # obtain memoryview of destination
-        dest = memoryview(out)
 
-        # ensure source is 1D
-        if isinstance(buf, np.ndarray):
-            buf = buf.reshape(-1, order='A')
-            # try to match itemsize
-            dtype = 'u%s' % dest.itemsize
-            buf = buf.view(dtype=dtype)
+def ensure_ndarray_from_memory(o, flatten=True):
+    """Convenience function to obtain a numpy ndarray using memory exposed by object `o`,
+    ensuring that no memory copies are made, and that `o` is not an object array.
 
-        # try to copy via memoryview
-        dest[:] = buf
+    Parameters
+    ----------
+    o : bytes-like
+        Any object exposing a memory buffer. On Python 3 this must be an object exposing
+        the new-style buffer interface. On Python 2 this can also be an object exposing
+        the old-style buffer interface.
+    flatten : bool, optional
+        If True, flatten any multi-dimensional inputs into a one-dimensional memoryview.
 
-    return out
+    Returns
+    -------
+    o : ndarray
 
+    """
 
-def ndarray_from_buffer(buf, dtype):
-    if isinstance(buf, np.ndarray):
-        arr = buf.reshape(-1, order='A').view(dtype)
-    else:
-        arr = np.frombuffer(buf, dtype=dtype)
-    return arr
+    if not isinstance(o, np.ndarray):
 
+        # first try to obtain a memoryview or buffer, needed to ensure that we don't
+        # subsequently copy memory when going via np.array()
 
-def ensure_text(l, encoding='utf-8'):
-    if isinstance(l, text_type):
-        return l
-    else:  # pragma: py3 no cover
-        return text_type(l, encoding=encoding)
+        if PY2:  # pragma: py3 no cover
+            # accept objects exposing either old-style or new-style buffer interface
+            try:
+                o = memoryview(o)
+            except TypeError:
+                o = buffer(o)
+
+        else:  # pragma: py2 no cover
+            o = memoryview(o)
+
+        # N.B., this is not documented, but np.array() will accept an object exposing
+        # a buffer interface, and will take a view of the memory rather than making a
+        # copy, preserving type information where present
+        o = np.array(o, copy=False)
+
+    # check for object arrays
+    if o.dtype == object:
+        raise ValueError('object arrays are not supported')
```

Comment:
This gets a bit tricky in practice. Not all things we want to use this with can enforce that it is `object` only. `Categorize` is a prime example. It'd be nice if we could separate this somehow so we can make use of this strategy more often.

Reply:
This is tricky. The way I've handled this currently is that this function is purely about obtaining a ndarray with contiguous memory, in the broader sense that all memory that needs to be encoded is in a single block (I've renamed the function to ``ensure_contiguous_ndarray()``). If a codec wants to be more flexible and accept object arrays also, then they can do that, but they need to handle that differently. Examples are ``Categorize``, ``MsgPack`` and ``JSON``.